l0_sampling/sketch.cpp

```diff
@@ -1,29 +1,74 @@
 #include "../include/sketch.h"
 #include <cassert>
+#include <iostream>
 
-Sketch::Sketch(vec_t n, long seed, double num_bucket_factor):
-    seed(seed), n(n), num_bucket_factor(num_bucket_factor) {
-  const unsigned num_buckets = bucket_gen(n, num_bucket_factor);
-  const unsigned num_guesses = guess_gen(n);
-  bucket_a = std::vector<vec_t>(num_buckets * num_guesses);
-  bucket_c = std::vector<vec_hash_t>(num_buckets * num_guesses);
+double Sketch::num_bucket_factor = 0.5;
+vec_t Sketch::n;
+size_t Sketch::num_elems;
+size_t Sketch::num_buckets;
+size_t Sketch::num_guesses;
+
+Sketch::SketchUniquePtr Sketch::makeSketch(const Sketch &old) {
+  return makeSketch(old.seed);
+}
+
+
+/*
+ * Static functions for creating sketches without a provided memory location
+ * these are useful for unit tests
+ */
+Sketch::SketchUniquePtr Sketch::makeSketch(long seed) {
+  void* loc = malloc(sketchSizeof());
+  return SketchUniquePtr(makeSketch(loc, seed), [](Sketch* s){ s->~Sketch(); free(s); });
 }
 
-Sketch::Sketch(const Sketch &old) : seed(old.seed), n(old.n),
-num_bucket_factor(old.num_bucket_factor) {
-  bucket_a = std::vector<vec_t>(old.bucket_a.size());
-  bucket_c = std::vector<vec_hash_t>(old.bucket_c.size());
+Sketch::SketchUniquePtr Sketch::makeSketch(long seed, std::fstream &binary_in) {
+  void* loc = malloc(sketchSizeof());
+  return SketchUniquePtr(makeSketch(loc, seed, binary_in), [](Sketch* s){ free(s); });
+}
+
+/*
+ * Static functions for creating sketches with a provided memory location.
+ * We use these in the production system to keep supernodes virtually contiguous.
+ */
+Sketch* Sketch::makeSketch(void* loc, long seed) {
```

Comment:
again, why are we returning the pointer to a location we provide?

Reply:
Here I actually think this is clearer. Just because these make functions are acting as constructors so returning a pointer to the 'new' object makes sense. Otherwise some of these return sketches and some don't which would be harder to understand.